tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -5822,6 +5832,560 @@ Status ConvertBatchMatMul(OpConverterParams* params) {
+  bool output_has_ellipsis;
+  EinsumHelper::ParseEquation(equation, &input_labels, &output_labels,
+                              &label_types, &input_label_counts,
+                              &output_label_counts, &input_has_ellipsis,
+                              &output_has_ellipsis);
+
+  if (input_has_ellipsis[0] || input_has_ellipsis[1] || output_has_ellipsis) {
+    // TODO(tfeher): Handle ellipsis like EinsumHelper::ProcessDimensions.
+    // Note: ProcessDimensions would introduce kBroadcasting labels, which we
+    // need to replace with kBatch before we call InitDescriptor.
+    VLOG(2) << "Ellipsis not yet supported";
+    return errors::Unimplemented("No conversion for einsum equation.");
+  }
+  if (absl::c_any_of(label_types, [](auto l) {
+        return l == EinsumHelper::DimensionType::kReduce ||
+               l == EinsumHelper::DimensionType::kBroadcasting;
+      })) {
+    VLOG(2) << "Einsum reductions not implemented";
+    return errors::Unimplemented("No conversion for einsum equation.");
+  }
+
+  auto no_duplicated_labels = [](const LabelCounts& label_counts) {
+    return absl::c_any_of(label_counts, [](int i) { return i > 1; });
+  };
+  if (no_duplicated_labels(input_label_counts[0]) ||
+      no_duplicated_labels(input_label_counts[1]) ||
+      no_duplicated_labels(output_label_counts)) {
+    VLOG(2) << "Einsum invalid label count";
+    return errors::Unimplemented("No conversion for einsum equation.");
+  }
+
+  if ((*input_a)->is_weights() && (*input_b)->is_tensor()) {
+    // We prefer to use FC layer, needs A as tensor and B as weight.
+    std::swap(*input_a, *input_b);
+    std::swap(input_labels[0], input_labels[1]);
+    std::swap(input_label_counts[0], input_label_counts[1]);
+  }
+
+  TF_RETURN_IF_ERROR(descriptor_a->InitDescriptor(
+      **input_a, input_labels[0], label_types, EinsumLayout::BFC));
+  TF_RETURN_IF_ERROR(
+      descriptor_b->InitDescriptor(**input_b, input_labels[1], label_types,
+                                   EinsumLayout::BCF, descriptor_a));
+  // TODO(tfeher): Update the permutation in the descriptors to avoid final
+  // transpose (if possible). Consider swapping the input if it eliminates
+  // final transpose.
+
+  // Get final transpose.
+  Labels matmul_output_labels(descriptor_a->b + descriptor_a->f +
+                              descriptor_b->f);
+  AssembleOutput(descriptor_a->permuted_labels.begin(),
+                 descriptor_b->permuted_labels.begin(), *descriptor_a,
+                 *descriptor_b, matmul_output_labels.begin());
+  TF_RETURN_IF_ERROR(FindIndices(output_labels.begin(), output_labels.end(),
+                                 matmul_output_labels.begin(),
+                                 matmul_output_labels.end(), final_transpose));
+  // Clear identity traspose.
+  std::vector<int> iota(final_transpose->size());
+  std::iota(iota.begin(), iota.end(), 0);
+  if (*final_transpose == iota) {
```

Comment:
I think it is more efficient to use a loop to check this.

Reply:
Replaced with a loop.